### PR TITLE
CI: Added PProf report each five minutes.

### PR DESCRIPTION
**Pull request by eloycoto**
To measure the usage of the CPU in a different parts of the test run the 
pprof in background each 5 minutes.                                      

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5078)

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

File: test/helpers/kubectl.go [hunk 1]
```diff
@@ -412,6 +412,57 @@ func (kub *Kubectl) BackgroundReport(commands ...string) (context.CancelFunc, er
 	return cancel, nil
 }
 
+// PprofReport runs pprof on cilium nodes each 5 minutes and saves the data
+// into the test-output.log
+func (kub *Kubectl) PprofReport() {
```
Comment (ianvernon) [inline test/helpers/kubectl.go]: Can you create an issue which tracks doing this for the runtime test suite as well?

File: test/helpers/kubectl.go [hunk 2]
```diff
@@ -412,6 +412,57 @@ func (kub *Kubectl) BackgroundReport(commands ...string) (context.CancelFunc, er
 	return cancel, nil
 }
 
+// PprofReport runs pprof on cilium nodes each 5 minutes and saves the data
+// into the test-output.log
```
Comment (ianvernon) [inline test/helpers/kubectl.go]: This documentation is inaccurate. Each dump of `gops pprof-cpu 1` is saved into its own file per-pod.

File: test/helpers/kubectl.go [hunk 3]
```diff
@@ -412,6 +412,57 @@ func (kub *Kubectl) BackgroundReport(commands ...string) (context.CancelFunc, er
...
+		if !res.WasSuccessful() {
+			return
+		}
+		files := kub.ExecPodCmd(KubeSystemNamespace, pod, `ls -1 /tmp/`)
+		for _, file := range files.ByLines() {
+			if !strings.Contains(file, "profile") {
+				continue
+			}
+
+			dest := filepath.Join(BasePath, testPath, fmt.Sprintf("%s-profile-%s", pod, file))
```
Comment (ianvernon) [inline test/helpers/kubectl.go]: Nit: append `.prof` to these filenames.

**Review by ianvernon — CHANGES_REQUESTED** [review]

**Comment by eloycoto** [comment]
test-me-please

File: test/helpers/cilium.go [hunk 4]
```diff
@@ -658,6 +658,42 @@ func (s *SSHMeta) CheckLogsForDeadlock() {
...
+
+			testPath, err := CreateReportDirectory()
+			if err != nil {
+				log.WithError(err).Errorf("cannot create test result path '%s'", testPath)
+				return
+			}
+
+			res := s.Exec("gops pprof-cpu 1")
+			if !res.WasSuccessful() {
+				return
```
Comment (ianvernon) [inline test/helpers/cilium.go]: Log that this failed.

**Comment by ianvernon** [comment]
@eloycoto please confirm that pprof for runtime tests works. I went through the archive on Jenkins and none of the runtime tests contain the pprof report. I had to check the `test-output.log` of some of the tests to see why:

From: `RuntimeMonitorTest_With_Sample_Containers_cilium_monitor_check_--to/test-output.log`

```
time="2018-08-03T07:39:42Z" level=debug msg="running command: gops pprof-cpu 1"
cmd: "gops pprof-cpu 1" exitCode: 1 stdout:

stderr:
Couldn't resolve addr or pid 1 to TCPAddress: couldn't get port for PID 1: open /home/vagrant/.config/gops/1: no such file or directory
```

With PRs like this, evidence should be provided that the expected functionality works as expected if it is not testable or evident by the tests themselves passing. This should be done by default for all PRs, so that we do not merge code that is not functioning correctly.

**Review by ianvernon — CHANGES_REQUESTED** [review]
One nit as a comment. My main feedback is located here:
https://github.com/cilium/cilium/pull/5078#issuecomment-410327310

**Comment by tgraf** [comment]
Do we need to run with `--pprof` to enable it in the DaemonSet?

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
comments addressed.

**Comment by eloycoto** [comment]
Provision error: 
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5465/

**Comment by eloycoto** [comment]
test-me-please

**Comment by ianvernon** [comment]
@eloycoto my primary concern [here](https://github.com/cilium/cilium/pull/5078#issuecomment-410327310) was not addressed:

> With PRs like this, evidence should be provided that the expected functionality works as expected if it is not testable or evident by the tests themselves passing. This should be done by default for all PRs, so that we do not merge code that is not functioning correctly.

I will check this manually when the tests finish, but proof that something which is added which doesn't have testing provided around it functions correctly should be provided as part of the PR as a comment or an update to the PR description itself.

**Comment by tgraf** [comment]
The following pprof files were created:
```
find . | grep -i prof
./RuntimePolicies_Checks_CIDR_L3_Policy/profile145008820.pprof
./RuntimeConnectivityTest_Basic_Connectivity_test_Test_connectivity_between_containers_with_policy_imported/profile564986294.pprof
./RuntimeConntrackTest_Conntrack-related_configuration_options_for_endpoints/profile718695783.pprof
./RuntimePolicies_Init_Policy_Test_Init_Egress_Policy_Test/profile092977890.pprof
./RuntimePolicies_Tests_Egress_To_World/profile656499881.pprof
./RuntimePolicyEnforcement_Policy_Enforcement_Default_Default_to_Always_with_policy/profile353300250.pprof
./111-/K8sPolicyTest_GuestBook_Examples_checks_policy_example/cilium-v4z5b-profile-profile857365234.pprof
./111-/K8sPolicyTest_GuestBook_Examples_checks_policy_example/cilium-lzd97-profile-profile422990380.pprof
./111-/K8sServicesTest_External_services_To_Services_first_policy,_match_service_by_labels/cilium-4ctwr-profile-profile055017092.pprof
./111-/K8sServicesTest_External_services_To_Services_first_policy,_match_service_by_labels/cilium-9lhj4-profile-profile707953002.pprof
./111-/K8sPolicyTest_Basic_Test_Denies_traffic_with_k8s_default-deny_egress_policy/cilium-v4z5b-profile-profile546951573.pprof
./111-/K8sPolicyTest_Basic_Test_Denies_traffic_with_k8s_default-deny_egress_policy/cilium-lzd97-profile-profile361188474.pprof
./111-/K8sPolicyTest_Namespaces_policies_Cilium_Network_policy_using_namespace_label_and_L7/cilium-lzd97-profile-profile622975492.pprof
./111-/K8sPolicyTest_Namespaces_policies_Cilium_Network_policy_using_namespace_label_and_L7/cilium-v4z5b-profile-profile485876406.pprof
./111-/K8sServicesTest_Checks_ClusterIP_Connectivity_Checks_service_on_same_node/cilium-9lhj4-profile-profile804674418.pprof
./111-/K8sServicesTest_Checks_ClusterIP_Connectivity_Checks_service_on_same_node/cilium-4ctwr-profile-profile951043989.pprof
./111-/K8sServicesTest_External_services_To_Services_first_endpoint_creation/cilium-4ctwr-profile-profile691151692.pprof
./111-/K8sServicesTest_External_services_To_Services_first_endpoint_creation/cilium-9lhj4-profile-profile660305054.pprof
./111-/K8sServicesTest_Bookinfo_Demo_Tests_bookinfo_demo/cilium-9lhj4-profile-profile010103167.pprof
./111-/K8sServicesTest_Bookinfo_Demo_Tests_bookinfo_demo/cilium-4ctwr-profile-profile795872340.pprof
./111-/K8sServicesTest_External_services_To_Services_first_policy/cilium-4ctwr-profile-profile258119517.pprof
./111-/K8sServicesTest_External_services_To_Services_first_policy/cilium-9lhj4-profile-profile774555607.pprof
./111-/K8sPolicyTest_Basic_Test_Denies_traffic_with_cnp_default-deny_egress_policy/cilium-v4z5b-profile-profile385234368.pprof
./111-/K8sPolicyTest_Basic_Test_Denies_traffic_with_cnp_default-deny_egress_policy/cilium-lzd97-profile-profile012949181.pprof
./111-/K8sPolicyTest_Basic_Test_ServiceAccount_Based_Enforcement/cilium-v4z5b-profile-profile822024516.pprof
./111-/K8sPolicyTest_Basic_Test_ServiceAccount_Based_Enforcement/cilium-lzd97-profile-profile396024499.pprof
./111-/K8sUpdates_Updating_Cilium_stable_to_master/cilium-pkcqx-profile-profile696334697.pprof
./111-/K8sUpdates_Updating_Cilium_stable_to_master/cilium-m62h5-profile-profile365280999.pprof
./111-/K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies/cilium-v4z5b-profile-profile144775290.pprof
./111-/K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies/cilium-lzd97-profile-profile473542672.pprof
```

Profiling works:
```
$  go tool pprof ./111-/K8sUpdates_Updating_Cilium_stable_to_master/cilium-pkcqx-profile-profile696334697.pprof
Local symbolization failed for cilium-agent: open /usr/bin/cilium-agent: no such file or directory
Some binary filenames not available. Symbolization may be incomplete.
Try setting PPROF_BINARY_PATH to the search path for local binaries.
File: cilium-agent
Build ID: 0005099267d2e41ff76a0f87aec9f78d253d62cb
Type: cpu
Time: Aug 7, 2018 at 5:19pm (CEST)
Duration: 30s, Total samples = 110ms ( 0.37%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 110ms, 100% of 110ms total
Showing top 10 nodes out of 135
      flat  flat%   sum%        cum   cum%
      20ms 18.18% 18.18%       30ms 27.27%  encoding/json.(*Decoder).readValue
      20ms 18.18% 36.36%       20ms 18.18%  runtime.futex
      10ms  9.09% 45.45%       10ms  9.09%  encoding/json.simpleLetterEqualFold
      10ms  9.09% 54.55%       10ms  9.09%  path.Clean
      10ms  9.09% 63.64%       10ms  9.09%  runtime.(*gcSweepBuf).push
      10ms  9.09% 72.73%       20ms 18.18%  runtime.gentraceback
      10ms  9.09% 81.82%       10ms  9.09%  runtime.step
      10ms  9.09% 90.91%       10ms  9.09%  syscall.Syscall
      10ms  9.09%   100%       10ms  9.09%  time.NewTimer
         0     0%   100%       10ms  9.09%  bufio.(*Writer).Flush
```

**Comment by tgraf** [comment]
@ianvernon I'm merging this as this is already incredibly useful and provides amazing visibility.

@eloycoto Can you explain why the prof files are only generated for 1.11 and some of the runtime tests?

**Review by tgraf — APPROVED** [review]

**Comment by ianvernon** [comment]
@tgraf thanks for providing the output. The `pprof` files are only generated for some of the runtime tests because the creation of them is not done on a per-test basis, but is on a timed-trigger (every 5 minutes) throughout the test suite.

**Pull request merged**